### PR TITLE
Allow lira to use cromwell-as-a-service

### PR DESCRIPTION
**Pull request by samanehsan**
Use cromwell-as-a-service to run workflows if "use_caas" and a "caas_key" are specified in the config file.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -5,5 +5,6 @@ google-cloud-storage
 flask
 gunicorn
 requests
-git+git://github.com/broadinstitute/cromwell-tools.git
+oauth2client
+git+git://github.com/broadinstitute/cromwell-tools.git@a939745db83fd4b5c72430104592ec39c11cb600
```
Comment (samanehsan) [inline requirements.txt]: I temporarily changed this to the commit from the cromwell-tools PR that includes using caas for testing. I'll change it back before merging!

**Comment by samanehsan** [comment]
jenkins retest

File: kubernetes/listener-deployment-setup.sh [hunk 2]
```diff
@@ -15,7 +14,13 @@ SS2_PREFIX=$6
 TENX_PREFIX=$7
 LOG_SINK_NAME=$8
 LOG_DESTINATION=$9
+USE_CAAS=${10}
```
Comment (rexwangcc) [inline kubernetes/listener-deployment-setup.sh:17]: Why is `${10}` here rather `$10`?

File: kubernetes/listener-deployment-setup.sh [hunk 3]
```diff
@@ -15,7 +14,13 @@ SS2_PREFIX=$6
...
+COLLECTION_NAME=${11:-""}
 VAULT_TOKEN_FILE=${VAULT_TOKEN_FILE:-"$HOME/.vault-token"}
+CROMWELL_URL=https://cromwell.mint-$ENV.broadinstitute.org/api/workflows/v1
+if ${USE_CAAS}; then
+    CROMWELL_URL=https://cromwell.caas-dev.broadinstitute.org/api/workflows/v1
```
Comment (rexwangcc) [inline kubernetes/listener-deployment-setup.sh:22]: So if we use CAAS here, currently the `caas-dev` is the only option for us?

File: kubernetes/listener-deployment.sh [hunk 4]
```diff
@@ -10,7 +10,12 @@ CERT_DIR=$3
 FULLCHAIN_FILENAME=$4
 PRIVKEY_FILENAME=$5
 DNS_ZONE_NAME=$6
-DOCKER_TAG=$7 #version of the quay.io/humancellatlas/secondary-analysis-lira image to deploy
+DOCKER_TAG=$7 # version of the quay.io/humancellatlas/secondary-analysis-lira image to deploy
+GCLOUD_PROJECT=$8
+USE_CAAS=$9  # whether or not to use cromwell-as-a-service
+WORKFLOW_COLLECTION_ID=${10}  # The name of the workflow-collection to create in SAM (e.g. "dev-workflows")
```
Comment (rexwangcc) [inline kubernetes/listener-deployment.sh:16]: What is the difference between `$10` and `${10}`?

File: kubernetes/listener-deployment.sh [hunk 5]
```diff
@@ -110,13 +124,21 @@ aws route53 change-resource-record-sets \
...
 
+
 # Set listener config secret
 printf "\nset listener config secret"
-LIRA_CONFIG_PATH=$CERT_DIR/config.json
-CONFIG_SECRET_NAME=$(bash populate-listener-config-secret.sh $LIRA_CONFIG_PATH)
+CONFIG_FILE=$CERT_DIR/config.json
+
+if $USE_CAAS; then
+    bash caas_login.py ${GCLOUD_PROJECT} ${ENV} ${CERT_DIR} ${WORKFLOW_COLLECTION_ID} ${FIRECLOUD_CONTACT_EMAIL} ${FIRECLOUD_GROUP_NAME}
```
Comment (rexwangcc) [inline kubernetes/listener-deployment.sh]: I guess `caas_login.py` here was a bash script?

File: kubernetes/populate-listener-config-secret.sh [hunk 6]
```diff
@@ -11,6 +12,8 @@ fi
...
 echo $config_secret_name
 
-kubectl create secret generic \
-    $config_secret_name \
-    --from-file=config=$config_file \
+if [ -z $caas_key_file ]; then
+    kubectl create secret generic $config_secret_name --from-file=config=$config_file
+else
+    kubectl create secret generic $config_secret_name --from-file=config=$config_file --from-file=caas_key=$caas_key_file
+fi
```
Comment (rexwangcc) [inline kubernetes/populate-listener-config-secret.sh]: nit: new line at the end

File: lira/api/notifications.py [hunk 7]
```diff
@@ -51,15 +52,26 @@ def post(body):
         }
         status_code = 201
     else:
+        # Get Cromwell auth
+        if lira_config.use_caas:
+            auth = {
+                'caas_key': os.environ.get('caas_key'),
```
Comment (rexwangcc) [inline lira/api/notifications.py]: I'm wondering if we could move this to `lira.py`, to let Lira choose between Cromwell instances at the startup. As long as lira started, I don't think there is a chance we will change the Cromwell instance from a non-caas to a caas one? I feel like using`os.environ.get('')` in endpoint controllers could cause some potential problems. What do you think?

Comment (rexwangcc) [reply lira/api/notifications.py]: For example, maybe we could `os.environ.get('caas_key')` in the lira.py and store it to the lira_config, and then `lira_config.caas_key` here?

File: run_docker.sh [hunk 8]
```diff
@@ -21,15 +23,31 @@ fi
...
     $port
+else
+    docker run \
+        --name lira \
+        --publish $port:$port \
+        -v $config:$mounted_config:ro \
+        -e listener_config=$mounted_config \
+        lira:$tag \
+        $port
+fi
```
Comment (rexwangcc) [inline run_docker.sh]: nit: new line at the end

**Comment by rexwangcc** [comment]
Dropped a few comments. Generally this looks great to me!

File: kubernetes/listener-deployment-setup.sh (hunk 3 above)
Comment (samanehsan) [reply kubernetes/listener-deployment-setup.sh:22]: Yes, caas is only available in dev at the moment

File: kubernetes/listener-deployment-setup.sh (hunk 2 above)
Comment (samanehsan) [reply kubernetes/listener-deployment-setup.sh:17]: This is necessary for positional parameters with more than one digit

File: kubernetes/listener-deployment.sh (hunk 5 above)
Comment (samanehsan) [reply kubernetes/listener-deployment.sh]: Good catch! This should be `caas_login.sh`: https://github.com/HumanCellAtlas/lira/blob/master/kubernetes/caas_login.sh

**Comment by samanehsan** [comment]
@rexwangcc I updated this PR to pass in CAAS-specific options to the workflows. Could you please take another look?

**Review by rexwangcc — APPROVED** [review]
LGTM!

**Review by rexwangcc — APPROVED** [review]
👍

**Pull request merged**